In [1]:
import pandas as pd
from pathlib import Path
from tabulate import tabulate

PROJECT_ROOT = Path.cwd()
while not (PROJECT_ROOT / ".git").exists() and PROJECT_ROOT != PROJECT_ROOT.parent:
    PROJECT_ROOT = PROJECT_ROOT.parent

df = pd.read_json(PROJECT_ROOT / "datasets" / "ViNumQA" / "test.json")
df.sample(n=5)

,pre_text,table,post_text,id,qa
366,[thuyết minh báo cáo tài chính hợp nhất năm 20...,"[[, 2010, 2009, 2008], [chi phí trả thưởng dựa...",[quyền chọn mua cổ phiếu các quyền chọn mua cổ...,GPN/2010/page_87.pdf-2,{'question': 'tỷ lệ phần trăm thay đổi chi phí...
52,[chúng tôi sử dụng 2 phương pháp: (1) so sánh ...,"[[Phương pháp, Giá trị, Tỷ trọng, Giá trị theo...",[.],masvn/2020/2020040-PPC_Báocáol_n__u-05022020/p...,{'question': 'Tổng giá trị theo tỷ trọng của c...
295,"[sử dụng phương pháp p/b và rnav để định giá, ...","[[Năm tài chính (31/12), 2018, 2019, 2020 (Dự ...",[.],masvn/2020/2020193-VN_IndustrialRealEstate_Upd...,{'question': 'Tốc độ tăng trưởng lợi nhuận sau...
224,[tập đoàn dollar general và các công ty con th...,"[[cải tạo đất đai, 20], [nhà cửa, 39-40], [nội...",[việc cải tạo tài sản thuê được phân bổ trong ...,DG/2009/page_77.pdf-1,{'question': 'Tỷ lệ khấu hao hàng năm cho các ...
170,"[doanh thu năm 2019 đạt 3,298 tỷ (+14 % yoy) v...","[[(Tỷ đồng), FY 2015, FY 2016, FY 2017, FY 201...",[.],masvn/2020/2020060-TLG_Companynote_MAS03.06.20...,{'question': 'Tính tỷ lệ tăng trưởng doanh thu...


In [2]:
len(df)

497

In [3]:
def formatting_pre_text(sample):
    return "\n".join(sample["pre_text"])

def formatting_table(sample):
    return tabulate(sample["table"][1:], headers=sample["table"][0], tablefmt="github")

def formatting_post_text(sample):
    return "\n".join(sample["post_text"])

def processing_input_question(sample):
    return sample["qa"]["question"]

def processing_program_content(sample):
    return sample["qa"]["program"]

def processing_answer_content(sample):
    return sample["qa"]["exe_ans"]

df["pre_text_processed"] = df.apply(lambda x: formatting_pre_text(x), axis=1)
df["post_text_processed"] = df.apply(lambda x: formatting_post_text(x), axis=1)
df["table_processed"] = df.apply(lambda x: formatting_table(x), axis=1)
df["input_question"] = df.apply(lambda x: processing_input_question(x), axis=1)
df["program_processed"] = df.apply(lambda x: processing_program_content(x), axis=1)
df["answer_processed"] = df.apply(lambda x: processing_answer_content(x), axis=1)
df.sample(n=5)

,pre_text,table,post_text,id,qa,pre_text_processed,post_text_processed,table_processed,input_question,program_processed,answer_processed
160,"[định giá và khuyến nghị, chúng tôi duy trì kh...","[[FY (tỷ đồng), FY18, FY19, FY20, FY21, FY22],...",[.],masvn/2020/2020163-VN_DHG_Update_BUY_MAS_20201...,{'question': 'Tính tổng doanh thu dự kiến năm ...,định giá và khuyến nghị\nchúng tôi duy trì khu...,.,| FY (tỷ đồng) | FY18 | FY19 ...,Tính tổng doanh thu dự kiến năm 2020 và 2021 c...,"add(1506, 1490)",2996.0
345,[.],"[[, 1Q18, 1Q19, YoY (%), Kế hoạch 2019 (%)], [...",[doanh thu tăng trưởng trong quý đầu năm có ph...,masvn/2020/2020013-VN_18042019MWGShortNote_TV_...,{'question': 'Biên lợi nhuận ròng cho quý 1 nă...,.,doanh thu tăng trưởng trong quý đầu năm có phầ...,| | 1Q18 | 1Q19 | YoY (%...,Biên lợi nhuận ròng cho quý 1 năm 2019 là bao ...,"divide(1041, 25017)",0.04161
179,[mục 11. thù lao ban điều hành thông tin về th...,"[[Loại chương trình, cột a số lượng chứng khoá...",[(a) không bao gồm 1.534.294 cổ phiếu hạn chế ...,PKG/2013/page_88.pdf-1,{'question': 'Tính đến ngày 31 tháng 12 năm 20...,mục 11. thù lao ban điều hành thông tin về thù...,(a) không bao gồm 1.534.294 cổ phiếu hạn chế c...,| Loại chương trình ...,"Tính đến ngày 31 tháng 12 năm 2013, tỷ lệ của ...","divide(2140954, 151945)",14.09032
240,[chúng tôi sử dụng phương pháp so sánh để định...,"[[FY (Dec), FY 2015, FY 2016, FY 2017, FY 2018...",[.],masvn/2020/2020029-POW_FlashNote1H2019/page_1_QA8,{'question': 'Tỷ suất sinh lời trên vốn chủ sở...,chúng tôi sử dụng phương pháp so sánh để định ...,.,| FY (Dec) | FY 2015 | FY 2016 ...,Tỷ suất sinh lời trên vốn chủ sở hữu (ROE) tru...,"table_average(ROE (%), none)",7.225
365,"[tính đến ngày 31 tháng 12 năm 2015, các khoản...","[[2016, $ 5754], [2017, 5933], [2018, 6113], [...","[sau khi hoàn thành xây dựng vào năm 2013, chú...",ANET/2015/page_156.pdf-2,{'question': 'Phần hiện tại của giá trị hiện t...,"tính đến ngày 31 tháng 12 năm 2015, các khoản ...","sau khi hoàn thành xây dựng vào năm 2013, chún...",| 2016 ...,Phần hiện tại của giá trị hiện tại các nghĩa v...,"divide(1336, 42546)",0.0314


In [4]:
df = df[["pre_text_processed", "table_processed", "post_text_processed", "input_question", "program_processed", "answer_processed"]]
# Flat column list, not [[...]] -- double brackets create a MultiIndex that
# silently breaks boolean filtering on df["generated_program"] later.
df.columns = ["pre_text", "table", "post_text", "question", "program", "answer"]
df["generated_program"] = ""
df["calculated_program"] = ""
df

,pre_text,table,post_text,question,program,answer,generated_program,calculated_program
0,thuyết minh báo cáo tài chính hợp nhất ( tiếp ...,| các thành phần của ảnh hưởng lũy kế của việc...,.,Sự thay đổi trong thu nhập ròng từ hiệu ứng tí...,"add(30, 1)",31.0,,
1,định giá và khuyến nghị:\nchúng tôi khuyến ngh...,| Năm tài chính (31/12) | FY17 | FY1...,.,"Theo dự phóng, doanh thu và lợi nhuận ròng quý...","subtract(9829, 642)",9187.0,,
2,"sử dụng phương pháp p/b và rnav để định giá, c...",| Năm tài chính (31/12) | 2016 | 2017 | ...,.,IDC có bao nhiêu ha quỹ đất sẵn sàng cho thuê ...,"add(495, 398)",893.0,,
3,27/10/13 26/10/14 25/10/15 30/10/16 29/10/17 2...,| | 27/10/2013 |...,.,Tỷ suất lợi nhuận trên đầu tư (ROI) của Applie...,"subtract(96.67, 100), divide(#0, 100)",-0.0333,,
4,thông tin tài chính bổ sung hiệu suất cổ phiếu...,| | 12/26/08 | ...,218 báo cáo thường niên năm 2013 của goldman s...,tỷ lệ lợi nhuận tích lũy tổng cộng theo phần t...,"subtract(248.36, 100), divide(#0, 100)",1.4836,,
...,...,...,...,...,...,...,...,...
492,thuyết minh báo cáo tài chính hợp nhất năm 201...,| ...,trong tổng số lợi ích thuế chưa được ghi nhận ...,Tỷ lệ phần trăm lợi ích thuế chưa được công nh...,"divide(131.8, 148.8)",0.88575,,
493,định giá và khuyến nghị:\nchúng tôi khuyến ngh...,| Năm tài chính (31/12) | FY17 | FY1...,.,Doanh thu trung bình từ năm tài chính 2017 đến...,"add(29710, 32662), add(#0, 35374), divide(#1, 3)",32582.0,,
494,định giá và khuyến nghị:\nchúng tôi khuyến ngh...,| Năm tài chính (31/12) | FY17 | FY1...,.,Tính phần trăm thay đổi EPS từ năm 2020 đến 2021.,"subtract(962, 788), divide(#0, 788)",0.22081,,
495,"trong quá trình kinh doanh thông thường, dựa t...",| ( đơn vị: nghìn ) | diện tích ròng chưa ph...,( a ) một giếng khoan thăm dò được lên kế hoạc...,Tỷ lệ phần trăm diện tích đất chưa phát triển ...,"divide(145, 586)",0.24744,,


In [5]:
SYSTEM_MESSAGE = """You are a financial analysis AI. Your task is to generate a sequential computation program to answer the question, based on the provided context.
 
### LIST OF 10 VALID OPERATORS:
 
1. add(a, b) -> a + b
2. subtract(a, b) -> a - b
3. multiply(a, b) -> a * b
4. divide(a, b) -> a / b
5. exp(a, b) -> a^b
6. greater(a, b) -> 1.0 if a > b, else 0.0
7. table_sum(val1, val2, val3, ...) -> sum of the values
8. table_average(val1, val2, val3, ...) -> arithmetic mean of the values
9. table_max(val1, val2, val3, ...) -> maximum of the values
10. table_min(val1, val2, val3, ...) -> minimum of the values
 
### RULES:
- Do not use free-form mathematical symbols ("+", "-", "*", "/") outside of parentheses. Every calculation must use one of the 10 operators above.
- Do not use square brackets `[]` inside table-type functions (e.g. write table_max(1, 2, 3), not table_max([1, 2, 3])).
- Do not perform mental calculations or provide explanations. The output must contain only the program string.
- Reference the result of a previous step using #0 (step 1), #1 (step 2), etc. Steps are separated by commas.
- Preserve the original number format from the context. If a value is missing, use 'none'."""

USER_MESSAGE_FRAME = """### CONTEXT:
[TEXT BEFORE TABLE]
{pre_text}
 
[TABLE]
{table}
 
[TEXT AFTER TABLE]
{post_text}
 
### QUESTION:
{question}
 
### PROGRAM:"""

## Batch API generation (0-shot, gpt-5-nano)

Uses OpenAI's Batch API (`/v1/responses`) instead of a synchronous per-row loop:
50% lower cost than real-time calls, in exchange for async turnaround (usually
minutes, up to 24h). This is a good fit here since we're scoring the full
497-sample test set offline rather than serving live requests.

Reasoning effort is `medium`: with no in-context examples, the model needs more
deliberation to get the operator syntax and table grounding right (0-shot has
nothing to imitate). See the few-shot batch notebook for the `low`-effort variant.

In [6]:
import os
from dotenv import load_dotenv
from openai import OpenAI

load_dotenv()

# Requires a real OpenAI API key (not the FPT Cloud / third-party BASE_URL used
# by the other notebooks) -- Batch API is an OpenAI-account-level feature.
OPENAI_API_KEY = os.environ["OPENAI_API_KEY"]
client = OpenAI(api_key=OPENAI_API_KEY)

MODEL = "gpt-5-nano"
REASONING_EFFORT = "medium"  # 0-shot: no examples to imitate, needs more deliberation

In [7]:
import json
from pathlib import Path

BATCH_DIR = Path(f"outputs/0shot_{MODEL}_batch")
BATCH_DIR.mkdir(parents=True, exist_ok=True)

REQUESTS_PATH = BATCH_DIR / "requests.jsonl"

def build_request(df_index, values):
    user_msg = USER_MESSAGE_FRAME.format(
        pre_text=values["pre_text"],
        table=values["table"],
        post_text=values["post_text"],
        question=values["question"],
    )
    return {
        "custom_id": str(df_index),
        "method": "POST",
        "url": "/v1/responses",
        "body": {
            "model": MODEL,
            "instructions": SYSTEM_MESSAGE,
            "input": user_msg,
            "reasoning": {"effort": REASONING_EFFORT},
            "max_output_tokens": 2048,
        },
    }

with open(REQUESTS_PATH, "w", encoding="utf-8") as f:
    for df_index, values in df.iterrows():
        f.write(json.dumps(build_request(df_index, values), ensure_ascii=False) + "\n")

print(f"Wrote {len(df)} requests to {REQUESTS_PATH}")

Wrote 497 requests to outputs\0shot_gpt-5-nano_batch\requests.jsonl


In [8]:
# Submit the batch job. Re-run-safe: if a batch_id was already saved from a
# previous run of this notebook, reuse it instead of submitting a duplicate job.
BATCH_ID_PATH = BATCH_DIR / "batch_id.txt"

if BATCH_ID_PATH.exists():
    batch_id = BATCH_ID_PATH.read_text().strip()
    print(f"Reusing existing batch job: {batch_id}")
else:
    batch_input_file = client.files.create(
        file=open(REQUESTS_PATH, "rb"),
        purpose="batch",
    )
    batch = client.batches.create(
        input_file_id=batch_input_file.id,
        endpoint="/v1/responses",
        completion_window="24h",
        metadata={"description": f"ViNumQA 0-shot {MODEL} ({REASONING_EFFORT} effort)"},
    )
    batch_id = batch.id
    BATCH_ID_PATH.write_text(batch_id)
    print(f"Submitted batch job: {batch_id}")

Submitted batch job: batch_6a5efa986f3c8190b202d84f72762508


In [9]:
# Poll until the batch completes. Safe to re-run this cell repeatedly --
# it just checks status and returns immediately if not done yet.
import time

TERMINAL_STATUSES = {"completed", "failed", "expired", "cancelled"}

def poll_batch(batch_id, poll_interval_s=30, max_wait_s=None):
    waited = 0
    while True:
        batch = client.batches.retrieve(batch_id)
        counts = batch.request_counts
        print(f"status={batch.status}  completed={counts.completed}/{counts.total}  failed={counts.failed}")
        if batch.status in TERMINAL_STATUSES:
            return batch
        if max_wait_s is not None and waited >= max_wait_s:
            print(f"Stopped polling after {waited}s (still {batch.status}). Re-run this cell later to keep waiting.")
            return batch
        time.sleep(poll_interval_s)
        waited += poll_interval_s

# max_wait_s=None polls indefinitely; set e.g. max_wait_s=600 to check in, then
# stop and resume later (batch jobs can take minutes to ~24h).
batch = poll_batch(batch_id, poll_interval_s=30, max_wait_s=None)

status=validating  completed=0/0  failed=0
status=validating  completed=0/0  failed=0
status=validating  completed=0/0  failed=0
status=in_progress  completed=0/497  failed=0
status=in_progress  completed=27/497  failed=0
status=in_progress  completed=62/497  failed=0
status=in_progress  completed=96/497  failed=0
status=in_progress  completed=132/497  failed=0
status=in_progress  completed=166/497  failed=0
status=in_progress  completed=203/497  failed=0
status=in_progress  completed=225/497  failed=0
status=in_progress  completed=260/497  failed=0
status=in_progress  completed=295/497  failed=0
status=in_progress  completed=321/497  failed=0
status=in_progress  completed=354/497  failed=0
status=in_progress  completed=363/497  failed=0
status=in_progress  completed=427/497  failed=0
status=in_progress  completed=459/497  failed=0
status=in_progress  completed=476/497  failed=0
status=in_progress  completed=476/497  failed=0
status=in_progress  completed=476/497  failed=0
status=in_pr

In [10]:
# Download and parse the batch output, mapping results back onto df via custom_id.
if batch.status != "completed":
    raise RuntimeError(f"Batch did not complete successfully: status={batch.status}, errors={batch.errors}")

output_file = client.files.content(batch.output_file_id)
OUTPUT_PATH = BATCH_DIR / "results.jsonl"
output_file.write_to_file(OUTPUT_PATH)

n_ok, n_err = 0, 0
with open(OUTPUT_PATH, "r", encoding="utf-8") as f:
    for line in f:
        record = json.loads(line)
        df_index = int(record["custom_id"])
        response = record.get("response")
        if response is None or response.get("status_code") != 200:
            n_err += 1
            continue
        body = response["body"]
        output_text = body.get("output_text")
        if output_text is None:
            # Fall back to walking the output items if output_text is absent.
            texts = []
            for item in body.get("output", []):
                for content in item.get("content", []):
                    if content.get("type") == "output_text":
                        texts.append(content["text"])
            output_text = "".join(texts)
        df.at[df_index, "generated_program"] = output_text.strip().strip("\n")
        n_ok += 1

print(f"Parsed {n_ok} successful / {n_err} failed responses out of {n_ok + n_err}.")

Parsed 497 successful / 0 failed responses out of 497.


In [11]:
"""
Parser + evaluation utilities for FinQA/VLSP-2025 style computation programs.

Program string format (as seen in the dataset / notebook), e.g.:
    "add(30, 1)"
    "subtract(9829, 642), divide(#0, 642)"
    "table_sum(-54, 30, 1), table_sum(-23, 292)"

- Steps are separated by top-level commas (commas NOT inside parentheses).
- Each step is `operator(arg1, arg2, ...)`.
- An arg can be a literal number, or a reference to a previous step's
  result using `#0`, `#1`, ... (0-indexed).
- `none` is treated as a missing/undefined value.

Two metrics are provided, matching the paper's definitions:
- Program Accuracy (PA): structural match between the generated program
  and the gold program (operator sequence + normalized numeric args).
- Execution Accuracy (EA): whether executing the generated program
  produces the same final numeric value as the gold answer.

NOTE on PA: FinQA-style datasets often have multiple valid gold programs
for the same question (see `program_re` in the paper). This implementation
only compares against a single gold program string. If you have alternative
gold programs available, run `compute_program_accuracy` against each and
take the max.
"""

import re
from typing import List, Tuple, Union, Optional


VALID_OPERATORS = {
    "add", "subtract", "multiply", "divide", "exp", "greater",
    "table_sum", "table_average", "table_max", "table_min",
}

_BINARY_OPS = {"add", "subtract", "multiply", "divide", "exp", "greater"}
_TABLE_OPS = {"table_sum", "table_average", "table_max", "table_min"}

_NUM_RE = re.compile(r"^-?\d+(\.\d+)?$")
_REF_RE = re.compile(r"^#(\d+)$")


def extract_program(raw_text: str) -> str:
    text = raw_text.strip()
    text = re.sub(r"```[a-zA-Z]*", "", text)
    text = text.replace("```", "").strip()

    op_names = "|".join(sorted(VALID_OPERATORS, key=len, reverse=True))
    pattern = rf"\b(?:{op_names})\([^()]*\)"
    matches = re.findall(pattern, text)

    if matches:
        return ", ".join(m.strip() for m in matches)
    return text


def _split_top_level(s: str, sep: str = ",") -> List[str]:
    parts = []
    depth = 0
    current = []
    for ch in s:
        if ch == "(":
            depth += 1
            current.append(ch)
        elif ch == ")":
            depth -= 1
            current.append(ch)
        elif ch == sep and depth == 0:
            parts.append("".join(current))
            current = []
        else:
            current.append(ch)
    if current:
        parts.append("".join(current))
    return [p.strip() for p in parts if p.strip() != ""]


def parse_program(program_str: str) -> List[Tuple[str, List[str]]]:
    program_str = program_str.strip().rstrip(",").strip()
    if not program_str:
        raise ValueError("Empty program string.")

    steps: List[Tuple[str, List[str]]] = []
    call_pattern = re.compile(r"\s*([a-zA-Z_]+)\(([^()]*)\)\s*")
    pos = 0
    text = program_str
    while pos < len(text):
        m = call_pattern.match(text, pos)
        if not m:
            raise ValueError(f"Malformed program near: '{text[pos:pos+30]}...'")
        op = m.group(1).strip()
        if op not in VALID_OPERATORS:
            raise ValueError(f"Unknown operator: '{op}'")
        args = _split_top_level(m.group(2), sep=",")
        steps.append((op, args))
        pos = m.end()
        if pos < len(text) and text[pos] == ",":
            pos += 1

    if not steps:
        raise ValueError("No valid steps parsed.")
    return steps


def _resolve_arg(arg: str, results: List[float]) -> Optional[float]:
    arg = arg.strip()

    ref_match = _REF_RE.match(arg)
    if ref_match:
        idx = int(ref_match.group(1))
        if idx >= len(results):
            raise ValueError(f"Reference #{idx} used before step {idx} was computed.")
        return results[idx]

    if arg.lower() == "none":
        return None

    cleaned = arg.replace(",", "").replace("%", "").replace("$", "").strip()
    if not _NUM_RE.match(cleaned):
        raise ValueError(f"Cannot parse numeric argument: '{arg}'")
    return float(cleaned)


def execute_program(program_str: str) -> Tuple[float, List[float]]:
    steps = parse_program(program_str)
    results: List[float] = []

    for op, raw_args in steps:
        vals = [_resolve_arg(a, results) for a in raw_args]

        if op in _BINARY_OPS:
            if len(vals) != 2:
                raise ValueError(f"Operator '{op}' expects 2 args, got {len(vals)}.")
            a, b = vals
            if a is None or b is None:
                raise ValueError(f"Operator '{op}' received a 'none' operand.")
            if op == "add":
                r = a + b
            elif op == "subtract":
                r = a - b
            elif op == "multiply":
                r = a * b
            elif op == "divide":
                if b == 0:
                    raise ValueError("Division by zero.")
                r = a / b
            elif op == "exp":
                r = a ** b
            elif op == "greater":
                r = 1.0 if a > b else 0.0

        elif op in _TABLE_OPS:
            if len(vals) < 1:
                raise ValueError(f"Operator '{op}' requires at least 1 arg.")
            if any(v is None for v in vals):
                raise ValueError(f"Operator '{op}' received a 'none' operand.")
            if op == "table_sum":
                r = sum(vals)
            elif op == "table_average":
                r = sum(vals) / len(vals)
            elif op == "table_max":
                r = max(vals)
            elif op == "table_min":
                r = min(vals)
        else:
            raise ValueError(f"Unknown operator: '{op}'")

        results.append(r)

    return results[-1], results


def _normalize_program(program_str: str) -> List[Tuple[str, Tuple[str, ...]]]:
    steps = parse_program(program_str)
    normalized = []
    for op, args in steps:
        norm_args = []
        for a in args:
            a = a.strip()
            if _REF_RE.match(a) or a.lower() == "none":
                norm_args.append(a.lower())
            else:
                cleaned = a.replace(",", "").replace("%", "").replace("$", "").strip()
                try:
                    norm_args.append(f"{round(float(cleaned), 6)}")
                except ValueError:
                    norm_args.append(a)
        normalized.append((op, tuple(norm_args)))
    return normalized


def compute_program_accuracy(
    generated_program: str,
    gold_program: str,
    alternative_gold_programs: Optional[List[str]] = None,
) -> float:
    try:
        gen_norm = _normalize_program(generated_program)
    except ValueError:
        return 0.0

    gold_candidates = [gold_program] + (alternative_gold_programs or [])
    for gold in gold_candidates:
        try:
            gold_norm = _normalize_program(gold)
        except ValueError:
            continue
        if gen_norm == gold_norm:
            return 1.0
    return 0.0


def compute_execution_accuracy(
    generated_program: str,
    gold_answer: Union[str, float],
    rel_tol: float = 1e-3,
    abs_tol: float = 1e-4,
) -> float:
    try:
        predicted, _ = execute_program(generated_program)
    except (ValueError, ZeroDivisionError, OverflowError):
        return 0.0

    try:
        gold = float(str(gold_answer).replace(",", "").replace("%", "").strip())
    except ValueError:
        return 0.0

    if abs(predicted - gold) <= max(abs_tol, rel_tol * abs(gold)):
        return 1.0
    return 0.0


def evaluate_dataframe(
    df,
    generated_col: str = "generated_program",
    gold_program_col: str = "program_processed",
    gold_answer_col: str = "answer_processed",
    alternative_gold_col: Optional[str] = None,
    extract_first: bool = True,
):
    df = df.copy()
    pa_scores = []
    ea_scores = []

    for _, row in df.iterrows():
        raw_generated = row[generated_col]
        generated = extract_program(raw_generated) if extract_first else raw_generated
        gold_program = row[gold_program_col]
        gold_answer = row[gold_answer_col]

        alt_programs = None
        if alternative_gold_col is not None and alternative_gold_col in row:
            alt_val = row[alternative_gold_col]
            if isinstance(alt_val, str) and alt_val.strip():
                alt_programs = [alt_val]

        pa = compute_program_accuracy(generated, gold_program, alt_programs)
        ea = compute_execution_accuracy(generated, gold_answer)

        pa_scores.append(pa)
        ea_scores.append(ea)

    df["pa_score"] = pa_scores
    df["ea_score"] = ea_scores

    summary = {
        "program_accuracy": sum(pa_scores) / len(pa_scores) if pa_scores else 0.0,
        "execution_accuracy": sum(ea_scores) / len(ea_scores) if ea_scores else 0.0,
    }
    return df, summary

In [12]:
df_scored, summary = evaluate_dataframe(
    df,
    generated_col="generated_program",
    gold_program_col="program",
    gold_answer_col="answer",
)

print(summary)

{'program_accuracy': 0.18309859154929578, 'execution_accuracy': 0.29577464788732394}


In [13]:
results_path = BATCH_DIR / "scored_results.csv"
summary_path = BATCH_DIR / "summary.json"

df_scored.to_csv(results_path, index=False)
with open(summary_path, "w", encoding="utf-8") as f:
    json.dump({"model": MODEL, "shot": "0-shot", "reasoning_effort": REASONING_EFFORT, "api": "batch", **summary}, f, ensure_ascii=False, indent=2)

print(f"Saved per-sample results to {results_path}")
print(f"Saved summary to {summary_path}")

Saved per-sample results to outputs\0shot_gpt-5-nano_batch\scored_results.csv
Saved summary to outputs\0shot_gpt-5-nano_batch\summary.json
